### PR TITLE
Loki: add missing interpolation to metric find query requests

### LINKED ISSUE #54907: Loki: add interpolation to the Query Variable Editor
We recently added a QueryVariableEditor for the Loki datasource, but it's missing interpolation in its `metricFindQuery` call.

Identified [here](https://github.com/grafana/grafana/pull/54625#pullrequestreview-1100695191)

### PR DESCRIPTION
**Pull request by matyax**
As identified in [this comment](https://github.com/grafana/grafana/pull/54625#pullrequestreview-1100695191), with a recent refactor we removed interpolation from the variable queries.

Added back in this PR, supporting the new format for `metricFindQuery` requests, while keeping legacy support for any request using the string format.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/54907

**Special notes for your reviewer**:

Create a Loki variable that uses another existing variable in any of the fields.
Expected result: the request should interpolate the variable.

### CODE REVIEW

**Comment by matyax** [comment]
Specifically adding @ivanahuckova to this PR so she can follow up on her comment. Any other review and comments are always welcome as well.

File: public/app/plugins/datasource/loki/datasource.ts [hunk 1]
```diff
@@ -305,16 +312,43 @@ export class LokiDatasource
...
+  async processMetricFindQuery(query: LokiVariableQuery) {
+    if (query.type === LokiVariableQueryType.LabelNames) {
+      return this.labelNamesQuery();
+    }
+
+    if (!query.label) {
+      return [];
+    }
+
+    // If we have query expr, use /series endpoint
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/datasource.ts:343]:
```suggestion
    // If we have stream selector, use /series endpoint
```

File: public/app/plugins/datasource/loki/datasource.test.ts [hunk 2]
```diff
@@ -639,22 +674,6 @@ describe('LokiDatasource', () => {
         it('then the correct label should be added for metrics query', () => {
           assertAdHocFilters('rate({bar="baz"}[5m])', 'rate({bar="baz", job="grafana"}[5m])', ds);
         });
-
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/datasource.test.ts:642]: Is there a reason why are these removed?

**Review by ivanahuckova — COMMENTED** [review]
Works great! Added some feedback below. Also have you considered adding tests specifically for variable queries with template variables?

File: public/app/plugins/datasource/loki/datasource.ts (hunk 1 above)
Comment (matyax) [reply public/app/plugins/datasource/loki/datasource.ts:343]: Done for both occurrences.

File: public/app/plugins/datasource/loki/datasource.ts [hunk 3]
```diff
@@ -305,16 +312,43 @@ export class LokiDatasource
...
-    return await this.processMetricFindQuery(interpolated);
+    if (typeof query === 'string') {
+      const interpolated = this.templateSrv.replace(query, {}, this.interpolateQueryExpr);
+      return await this.legacyProcessMetricFindQuery(interpolated);
+    }
+
+    const interpolatedQuery = {
+      ...query,
+      label: this.templateSrv.replace(query.label, {}, this.interpolateQueryExpr),
+      stream: this.templateSrv.replace(query.stream, {}, this.interpolateQueryExpr),
```
[the remark this answers is not in the document]
Comment (matyax) [reply public/app/plugins/datasource/loki/datasource.ts:328]: Makes sense. I'll update the other usage as well https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/datasource.ts#L313

File: public/app/plugins/datasource/loki/datasource.test.ts (hunk 2 above)
Comment (matyax) [reply public/app/plugins/datasource/loki/datasource.test.ts:642]: No, good catch, it was a bad merge.

Comment (matyax) [reply public/app/plugins/datasource/loki/datasource.test.ts]: Added back.

File: public/app/plugins/datasource/loki/datasource.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (matyax) [reply public/app/plugins/datasource/loki/datasource.ts]: Done.

**Comment by matyax** [comment]
@ivanahuckova 

> Also have you considered adding tests specifically for variable queries with template variables?

Yes. Given how it's implemented, the best thing I could come up with so far was by asserting that we call `templateSrv` in the expected places. Not a fan of this approach since we're testing implementation data, but given the current state of mocking and I/O nature of the methods we need to test, I think this is a far enough approach to ensure we're interpolating variables where we're supposed to.

File: public/app/plugins/datasource/loki/datasource.ts [hunk 4]
```diff
@@ -305,24 +312,51 @@ export class LokiDatasource
...
+    if (!query.label) {
+      return [];
+    }
+
+    // If we have stream selector, use /series endpoint
+    if (query.stream) {
+      return this.labelValuesSeriesQuery(query.stream, query.label);
+    }
+
+    return this.labelValuesQuery(query.label);
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/datasource.ts:348]: This is just a personal preference, but I would wrap this part with 
`if (query.type === LokiVariableQueryType.LabelValues) {}`

and then default to `return []` to make it more clear that these different situations happen when we have `LabelValues` query type. And for anything else (that should not happen), default to `[]`.

**Review by ivanahuckova — APPROVED** [review]
LGTM 🚀  Good job!

File: public/app/plugins/datasource/loki/datasource.ts (hunk 4 above)
Comment (matyax) [reply public/app/plugins/datasource/loki/datasource.ts:348]: Makes sense. The reason why I structured it like this is because, by `label` being possibly undefined, I could just check once and avoid any further `if undefined` tests or defaults.
